pkg/execution/state/redis_state/redis_state.go

```diff
@@ -380,59 +382,127 @@ func (m mgr) Load(ctx context.Context, id state.Identifier) (state.State, error)
 }
 
 func (m mgr) SaveResponse(ctx context.Context, i state.Identifier, r state.DriverResponse, attempt int) (state.State, error) {
+	var (
+		data            any
+		err             error
+		typ             enums.HistoryType
+		funcFailHistory state.History
+	)
+
+	now := time.Now()
+
 	if r.Err == nil {
-		// Save the output.
-		str, err := json.Marshal(r.Output)
-		if err != nil {
-			return nil, err
+		typ = enums.HistoryTypeStepCompleted
+		if data, err = json.Marshal(r.Output); err != nil {
+			return nil, fmt.Errorf("error marshalling step output: %w", err)
 		}
-		if err := m.r.HSet(ctx, m.kf.Actions(ctx, i), r.Step.ID, str).Err(); err != nil {
-			return nil, err
+	} else {
+		typ = enums.HistoryTypeStepErrored
+		data = r.Err.Error()
+		if r.Final() {
+			typ = enums.HistoryTypeStepFailed
+			funcFailHistory = state.History{
+				Type:       enums.HistoryTypeFunctionFailed,
+				Identifier: i,
+				CreatedAt:  now,
```

Comment:
Will this store the time in microseconds? Is this okay?

Reply:
This is encoded by the binary encoder which uses JSON marshalling with gzip under the hood, so we're okay.